src/mpi-base.jl

```diff
@@ -5,11 +5,73 @@ typealias MPIDatatype Union{Char,
 typealias MPIBuffertype{T} Union{Ptr{T}, Array{T}, Ref{T}}
 
 # Define a function mpitype(T) that returns the MPI datatype code for
-# a given type T. The dictonary is defined in __init__ so the module
-# can be precompiled
+# a given type T. In the case the the type does not exist, it is created and
+# then returned. The dictonary is defined in __init__ so the module can be
+# precompiled
+
+# accessor and creation function for getting MPI datatypes
+function mpitype{T}(::Type{T})
+
+    if haskey(mpitype_dict, T)  # if the datatype already exists
```

Comment:
What happens if one creates an MPI datatype for a Julia type that contains pointer? There is no corresponding MPI datatype.

MPI datatypes need to have a very specific layout that can be expressed in terms of offsets. I think only bitstypes fit this. If you need to send more complex object, you will need to serialize it. The MPI package already handles this.